Labelling fails if v-site handler added after library charge handler

**Describe the bug**

Labelling a molecule with a FF that contains a library charge handler that was added after a v-site handler fails

[This flag](https://github.com/openforcefield/openff-toolkit/blob/921fbe2d48f3069adc3af6e919349dee6deefa26/openff/toolkit/typing/engines/smirnoff/forcefield.py#L1412) should likely be inside the next loop.

**To Reproduce**

```
from openff.toolkit.typing.engines.smirnoff import ForceField
from openff.toolkit.typing.engines.smirnoff.parameters import (
    LibraryChargeHandler,
    VirtualSiteHandler
)
from simtk import unit as simtk_unit

force_field = ForceField()
force_field.get_parameter_handler("Electrostatics")

vsite_handler: VirtualSiteHandler = force_field.get_parameter_handler(
    "VirtualSites"
)
vsite_handler.add_parameter(
    {
        "smirks": "[#6:1][#9:2]",
        "name": "EP",
        "type": "BondCharge",
        "distance": 1.0 * simtk_unit.angstrom,
        "match": "once",
        "charge_increment1": 0.2 * simtk_unit.elementary_charge,
        "charge_increment2": 0.1 * simtk_unit.elementary_charge,
        "sigma": 1.0 * simtk_unit.angstrom,
        "epsilon": 0.0 * simtk_unit.kilocalorie_per_mole,
    }
)

library_handler: LibraryChargeHandler = force_field.get_parameter_handler(
    "LibraryCharges"
)
library_handler.add_parameter(
    {
        "smirks": "[F:2][C:1]([H:3])([H:4])([H:5])",
        "charge": [
            0.3 * simtk_unit.elementary_charge,
            -0.15 * simtk_unit.elementary_charge,
            -0.05 * simtk_unit.elementary_charge,
            -0.05 * simtk_unit.elementary_charge,
            -0.05 * simtk_unit.elementary_charge
        ]
    }
)

print(force_field.label_molecules(Molecule.from_smiles("CF").to_topology()))
```

**Output**

```
                if param_is_list:
                    for match in matches:
>                       parameter_matches[match] = [
                            m.parameter_type for m in matches[match]
                        ]
E                       TypeError: '_Match' object is not iterable

/xxx/openff/toolkit/typing/engines/smirnoff/forcefield.py:1435: TypeError
```

**Computing environment (please complete the following information):**
 - Operating system
 - Output of running `conda list` 

```
openff-toolkit-base       0.10.3             pyhd8ed1ab_0    conda-forge
```

**Additional context**
<!-- Add any other context about the problem here. -->

--- a/docs/releasehistory.md
+++ b/docs/releasehistory.md
@@ -15,6 +15,14 @@ Releases follow the `major.minor.micro` scheme recommended by [PEP440](https://w
 - [PR #1284](https://github.com/openforcefield/openforcefield/pull/1284): Fixes
   [Issue #1283](https://github.com/openforcefield/openff-toolkit/issues/1283) - force fields containing `BondCharge` virtual sites cannot be loaded due to an issue with how `outOfPlaneAngle` and `inPlaneAngle` keywords are validated
 
+### Minor bugfixes
+- [PR #1290](https://github.com/openforcefield/openforcefield/pull/1290): Fixes
+  [Issue #1216](https://github.com/openforcefield/openff-toolkit/issues/1216) by adding internal logic to handle
+  the possibility that multiple vsites share the same parent atom, and makes the return value of 
+  `VirtualSiteHandler.find_matches` be closer to the base class.
+
+
+
 ## 0.10.5 Bugfix release
 
 - [PR #1252](https://github.com/openforcefield/openforcefield/pull/1252): Refactors virtual 

--- a/openff/toolkit/tests/test_forcefield.py
+++ b/openff/toolkit/tests/test_forcefield.py
@@ -2181,6 +2181,49 @@ class TestForceField:
 
         assert hash(ff_with_id) == hash(ff_without_id)
 
+    def test_issue_1216(self):
+        """Test that the conflict between vsitehandler and librarychargehandler
+        identified in issue #1216 remains resolved.
+
+        https://github.com/openforcefield/openff-toolkit/issues/1216
+        """
+        force_field = ForceField()
+        force_field.get_parameter_handler("Electrostatics")
+
+        vsite_handler: VirtualSiteHandler = force_field.get_parameter_handler(
+            "VirtualSites"
+        )
+        vsite_handler.add_parameter(
+            {
+                "smirks": "[#6:1][#9:2]",
+                "name": "EP",
+                "type": "BondCharge",
+                "distance": 1.0 * unit.angstrom,
+                "match": "all_permutations",
+                "charge_increment1": 0.2 * unit.elementary_charge,
+                "charge_increment2": 0.1 * unit.elementary_charge,
+                "sigma": 1.0 * unit.angstrom,
+                "epsilon": 0.0 * unit.kilocalorie_per_mole,
+            }
+        )
+
+        library_handler: LibraryChargeHandler = force_field.get_parameter_handler(
+            "LibraryCharges"
+        )
+        library_handler.add_parameter(
+            {
+                "smirks": "[F:2][C:1]([H:3])([H:4])([H:5])",
+                "charge": [
+                    0.3 * unit.elementary_charge,
+                    -0.15 * unit.elementary_charge,
+                    -0.05 * unit.elementary_charge,
+                    -0.05 * unit.elementary_charge,
+                    -0.05 * unit.elementary_charge
+                ]
+            }
+        )
+        force_field.label_molecules(Molecule.from_smiles("CF").to_topology())
+
 
 def generate_monatomic_ions():
     return (

--- a/openff/toolkit/tests/test_forcefield.py
+++ b/openff/toolkit/tests/test_forcefield.py
@@ -2218,8 +2218,8 @@ class TestForceField:
                     -0.15 * unit.elementary_charge,
                     -0.05 * unit.elementary_charge,
                     -0.05 * unit.elementary_charge,
-                    -0.05 * unit.elementary_charge
-                ]
+                    -0.05 * unit.elementary_charge,
+                ],
             }
         )
         force_field.label_molecules(Molecule.from_smiles("CF").to_topology())

--- a/openff/toolkit/tests/test_parameters.py
+++ b/openff/toolkit/tests/test_parameters.py
@@ -2372,11 +2372,11 @@ class TestVirtualSiteHandler:
 
         matched_smirks = defaultdict(set)
 
-        for match in matches:
-
-            matched_smirks[match.environment_match.topology_atom_indices].add(
-                (match.parameter_type.smirks, match.parameter_type.name)
-            )
+        for match_list in matches.values():
+            for match in match_list:
+                matched_smirks[match.environment_match.topology_atom_indices].add(
+                    (match.parameter_type.smirks, match.parameter_type.name)
+                )
 
         assert {**matched_smirks} == expected_matches
 

--- a/openff/toolkit/typing/engines/smirnoff/forcefield.py
+++ b/openff/toolkit/typing/engines/smirnoff/forcefield.py
@@ -1493,7 +1493,6 @@ class ForceField:
                     param_is_list = True
 
                 matches = parameter_handler.find_matches(top_mol)
-                print(matches)
                 # Remove the chemical environment matches from the
                 # matched results.
 

--- a/openff/toolkit/typing/engines/smirnoff/forcefield.py
+++ b/openff/toolkit/typing/engines/smirnoff/forcefield.py
@@ -1486,8 +1486,8 @@ class ForceField:
         for molecule_idx, molecule in enumerate(topology.reference_molecules):
             top_mol = Topology.from_molecules([molecule])
             current_molecule_labels = dict()
-            param_is_list = False
             for tag, parameter_handler in self._parameter_handlers.items():
+                param_is_list = False
 
                 if type(parameter_handler) == VirtualSiteHandler:
                     param_is_list = True

--- a/openff/toolkit/typing/engines/smirnoff/parameters.py
+++ b/openff/toolkit/typing/engines/smirnoff/parameters.py
@@ -5151,7 +5151,7 @@ class VirtualSiteHandler(_NonbondedHandler):
         entity: Topology,
         transformed_dict_cls=dict,
         unique=False,
-    ) -> Dict[Tuple(int), List[ParameterHandler._Match]]:
+    ) -> Dict[Tuple[int], List[ParameterHandler._Match]]:
 
         assigned_matches_by_parent = self._find_matches_by_parent(entity)
         return_dict = {}

--- a/openff/toolkit/typing/engines/smirnoff/parameters.py
+++ b/openff/toolkit/typing/engines/smirnoff/parameters.py
@@ -5151,13 +5151,12 @@ class VirtualSiteHandler(_NonbondedHandler):
         entity: Topology,
         transformed_dict_cls=dict,
         unique=False,
-    ) -> List[ParameterHandler._Match]:
+    ) -> Dict[int, List[ParameterHandler._Match]]:
 
         assigned_matches_by_parent = self._find_matches_by_parent(entity)
-        assigned_matches = []
-
+        return_dict = {}
         for parent_index, assigned_parameters in assigned_matches_by_parent.items():
-
+            assigned_matches = []
             for assigned_parameter, match_orientations in assigned_parameters:
 
                 for match in match_orientations:
@@ -5165,8 +5164,9 @@ class VirtualSiteHandler(_NonbondedHandler):
                     assigned_matches.append(
                         ParameterHandler._Match(assigned_parameter, match)
                     )
+            return_dict[(parent_index,)] = assigned_matches
 
-        return assigned_matches
+        return return_dict
 
     def create_force(self, system: openmm.System, topology: Topology, **kwargs):
 

--- a/openff/toolkit/typing/engines/smirnoff/parameters.py
+++ b/openff/toolkit/typing/engines/smirnoff/parameters.py
@@ -54,7 +54,7 @@ import logging
 import re
 from collections import OrderedDict, defaultdict
 from enum import Enum
-from typing import Any, List, Optional, Union, cast
+from typing import Any, List, Optional, Tuple, Union, cast
 
 try:
     import openmm
@@ -5151,7 +5151,7 @@ class VirtualSiteHandler(_NonbondedHandler):
         entity: Topology,
         transformed_dict_cls=dict,
         unique=False,
-    ) -> Dict[int, List[ParameterHandler._Match]]:
+    ) -> Dict[Tuple(int), List[ParameterHandler._Match]]:
 
         assigned_matches_by_parent = self._find_matches_by_parent(entity)
         return_dict = {}

